src/tls13.c

```diff
@@ -4410,10 +4410,48 @@ int SendTls13ClientHello(WOLFSSL* ssl)
     }
 #endif
 
-    /* Include length of TLS extensions. */
-    ret = TLSX_GetRequestSize(ssl, client_hello, &args->length);
-    if (ret != 0)
-        return ret;
+    {
+#ifdef WOLFSSL_DTLS_CH_FRAG
+        int maxFrag = wolfSSL_GetMaxFragSize(ssl, MAX_RECORD_SIZE);
+        word16 lenWithoutExts = args->length;
+#endif
+
+        /* Include length of TLS extensions. */
+        ret = TLSX_GetRequestSize(ssl, client_hello, &args->length);
+        if (ret != 0)
+            return ret;
+
+#ifdef WOLFSSL_DTLS_CH_FRAG
+        if (ssl->options.dtls && args->length > maxFrag &&
+                TLSX_Find(ssl->extensions, TLSX_COOKIE) == NULL) {
+            /* Try again with an empty key share if we would be fragmenting
+             * without a cookie */
+            TLSX* ks = TLSX_Find(ssl->extensions, TLSX_KEY_SHARE);
+            if (ks == NULL) {
+                WOLFSSL_MSG("No key share and CH can't fit in one fragment.");
+                return BUFFER_ERROR;
+            }
+            args->length = lenWithoutExts;
+            if (ssl->dtls13KSE != NULL)
+                TLSX_KeyShare_FreeAll(ssl->dtls13KSE, ssl->heap);
+            ssl->dtls13KSE = (KeyShareEntry*)ks->data;
+            ks->data = NULL;
+            ret = TLSX_GetRequestSize(ssl, client_hello, &args->length);
+            if (ret != 0) {
+                /* Restore key share data */
+                ks->data = ssl->dtls13KSE;
+                ssl->dtls13KSE = NULL;
+                return ret;
+            }
+            if (args->length > maxFrag) {
+                WOLFSSL_MSG("Can't fit first CH in one fragment.");
+                return BUFFER_ERROR;
+            }
+            WOLFSSL_MSG("Sending empty key share so we don't fragment CH1");
+            ssl->options.dtlsSentEmptyKS = 1;
+        }
+#endif
+    }
```

Comment:
On second thought I think it would be better to just set `ssl->options.dtlsSentEmptyKS` here and gate actually writing the key shares inside `TLSX_KeyShare_*` functions. Thoughts @rizlik ?

Reply:
I would keep things way simpler by just emptying the keyshare extensions here. I would not add any specific flag.  After HRR we will re-generate only the one keyshare entry (the one selected by the server) and here we are saving all of them. We already support sending empty KeyShare in the first CH (wolfSSL_NoKeyShare) without using any extra flag. While this is a good optimization, I would leave it for a later moment.

Reply:
I think I agree. Clearing it makes the change much simpler. I was hoping to not have to generate the keys twice as it can be a time consuming op. Will leave this as a future optimization.